flock_agent/daemon/daemon.py

```diff
@@ -357,3 +361,94 @@ def log(self, request, response, time):
         # Change permissions of unix socket
         os.chmod(self.unix_socket_path, 0o660)
         os.chown(self.unix_socket_path, 0, self.gid)
+
+    async def autoupdate_loop(self):
+        # Autoupdate is only available for macOS right now; Linux uses package managers
+        if Platform.current() != Platform.MACOS:
+            return
+
+        while True:
+            self.c.log("Daemon", "autoupdate_loop", "Checking for updates")
+
+            try:
+                # Query github for the latest version of Flock Agent
+                r = requests.get(
+                    "https://api.github.com/repos/firstlookmedia/flock-agent/releases/latest"
+                )
+                release = r.json()
+                latest_version = release["tag_name"].lstrip("v")
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"installed version: {self.c.version}, latest version: {latest_version}",
+                )
+                if latest_version <= self.c.version:
+                    continue
+
+                # Find the pkg asset
+                url = None
+                filename = None
+                for asset in release["assets"]:
+                    if asset["name"].endswith(".pkg"):
+                        url = asset["browser_download_url"]
+                        filename = asset["name"]
+                        break
+
+                if not url:
+                    self.c.log("Daemon", "autoupdate_loop", "could not find .pkg file")
+                    continue
+
+                # Download the update
+                self.c.log("Daemon", "autoupdate_loop", f"downloading {url}")
+                r = requests.get(url)
+
+                os.makedirs(os.path.join(self.lib_dir, "updates"), exist_ok=True)
+                download_filename = os.path.join(self.lib_dir, "updates", filename)
+                with open(download_filename, "wb") as f:
+                    f.write(r.content)
+
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"download complete: {download_filename}",
+                )
+
+                # Verify that it's codesigned
+                p = subprocess.run(
+                    ["/usr/sbin/pkgutil", "--check-signature", download_filename],
+                    stdout=subprocess.PIPE,
+                )
+                if (
+                    p.returncode != 0
+                    or "Status: signed by a developer certificate issued by Apple for distribution"
+                    not in p.stdout.decode()
+                    or "Developer ID Installer: FIRST LOOK PRODUCTIONS"
```

Comment:
I'm not sure how apple signing works. Is this a name apple has to approve? Is it guaranteed to be unique? If yes to both, is there anything preventing me from getting the developer id "FIRST LOOK PRODUCTIONS HAPPY FUNTIME CORP?"

Reply:
Also is there anywhere else that attacker-generated text could appear? I'm worried about a similar attack to https://arstechnica.com/information-technology/2018/06/decades-old-pgp-bug-allowed-hackers-to-spoof-just-about-anyones-signature/

Reply:
If you run `pkgutil --check-signature` on an unsigned packaged, the return code is 1 and the output is:

```
$ /usr/sbin/pkgutil --check-signature FlockAgent-0.1.0.pkg
Package "FlockAgent-0.0.12.pkg":
   Status: no signature
```

If you run it on a valid package signed with our certificate, the return code is 0 and the output is:

```
$ /usr/sbin/pkgutil --check-signature FlockAgent-0.1.0.pkg                                       micahlee@trapdoor
Package "FlockAgent-0.1.0.pkg":
   Status: signed by a developer certificate issued by Apple for distribution
   Certificate Chain:
    1. Developer ID Installer: FIRST LOOK PRODUCTIONS, INC. (P24U45L8P5)
       SHA256 Fingerprint:
           46 E2 4B DE 4A 50 C3 73 19 90 D4 CD 61 20 C6 A2 D9 56 66 BF 0C 6E 
           AC 22 EC 57 58 FE 46 3C A1 3C
       ------------------------------------------------------------------------
    2. Developer ID Certification Authority
       SHA256 Fingerprint:
           7A FC 9D 01 A6 2F 03 A2 DE 96 37 93 6D 4A FE 68 09 0D 2D E1 8D 03 
           F2 9C 88 CF B0 B1 BA 63 58 7F
       ------------------------------------------------------------------------
    3. Apple Root CA
       SHA256 Fingerprint:
           B0 B1 73 0E CB C7 FF 45 05 14 2C 49 F1 29 5E 6E DA 6B CA ED 7E 2C 
           68 C5 BE 91 B5 A1 10 01 F0 24
```

In order to get a valid code-signing certificate, you need to make an Apple Developer account (and pay Apple $99/year), and they act as a sort of certificate authority. They verify your identity, and for a business account I think they require some sort of business documentation.

I could verify that it's signed by the exact signing key by checking the fingerprint, but I'm hesitant to do that because of how Apple code signing certificates work. You can revoke them and generate new ones, they expire, etc. So my plan was instead to just tie it to the company name. (I'm actually not sure what the `P24U45L8P5` part is, but it's not part of our company name.)

A potential improvement is I could match on the string `Developer ID Installer: FIRST LOOK PRODUCTIONS, INC. (` instead of `Developer ID Installer: FIRST LOOK PRODUCTIONS`.

And as far as attacker-generated text? I imagine an attacker that creates a malicious certificate could put names in the certificate chain that get displayed, and include these snippets of next that I'm searching for. However, for the return code to equal 0, `pkgutil` needs to verify that the certificate is actually signed by the trusted Apple certificate authority.

So I believe that the attacker couldn't just make a fake cert with these strings, they would have to actually convince Apple to sign their certificate. Only then will it return with return code 0.

Reply:
I'm guessing developer id should be consistent? But I'm not sure about that. What about matching `Developer ID Installer: FIRST LOOK PRODUCTIONS, INC. ([^)\n]+)\n`? I'm hoping Apple wouldn't allow two developers to have the same developer ID name, even if the 2nd managed to incorporate a company with the same name.

I'm not sure if someone incorporating a company with a similar name, getting an Apple developer ID based on that company, and getting control of the github repo is our our threat model.

Reply:
I guess the regex makes it so we can't simply do `if a in b`. I don't feel super strongly about matching this exactly.